### PR TITLE
XMPP server migration from metronome to Prosody 0.10

### PR DESCRIPTION
**Pull request by M5oul**
- [Ticket #67: Migrate from Metronome to Prosody](https://dev.yunohost.org/issues/67)
---

### LDAP authentication module
- [x] [conf] add module for LDAP authentication from [modules repository](https://hg.prosody.im/prosody-modules)
- [x] [ynh-dev] adapt [ynh-dev symlinks for prosody modules](https://github.com/YunoHost/ynh-dev/pull/23)

### Templates
- [x] [conf] remove metronome configuration templates
- [x] [conf] add prosody configuration templates
- [x] [conf] enable MUC and BOSH
- ~~[conf] try to have similar configuration as metronome: [$1](https://github.com/YunoHost/yunohost-config-metronome/commit/fb45f56dcf62c962bd0bb58602545dfa70f63754)~~

### Other
- [x] [conf] adapt regen-conf, backup and restore scripts
- [x] [conf] change `service` file.
- [x] [cert] #222 handle prosody permissions on certificates.

### Debian packages management
- [x] change `debian/control` to install Debian package of Prosody.
- [x] use meta-package to conflict and remove metronome package
- [ ] handle well the debian package migration
- [ ] install directly prosody 0.10 and not 0.9
- [x] Use [Prosody repository](https://prosody.im/download/package_repository)
- [x] add repo line on `/etc/apt/sources.list.d/yunohost.list` and GPG key
  - [x] add into the https://github.com/YunoHost/install_script/pull/45

### Migration framework
- [x] [migration] write a Python migration class with [migration framework](https://github.com/YunoHost/yunohost/pull/195)
  - [x] [migration] regen-conf for prosody service
  - [x] copy data `rosters`, `vcards` …etc from `/var/lib/metronome/` to `/var/lib/prosody`

---

### Linked PR in other repositories
- [x] https://github.com/YunoHost/ynh-dev/pull/23
- [x] https://github.com/YunoHost/install_script/pull/45
- [x] https://github.com/YunoHost/doc/issues/473
- [ ] [linter] [update service list](https://github.com/YunoHost/package_linter/blob/master/package_linter.py#L149)

---
### Set up a testing instance

Inside [`ynh-dev` ](https://github.com/YunoHost/ynh-dev/):

go on `prosody` branches into `ynh-dev` and `yunohost` git repositories:
```bash
cd /vagrant
git checkout -b prosody
cd /vagrant/yunohost
git checkout -b prosody
```

Use prosody branch of `yunohost` and unstable of `moulinette` git repositories:
```bash
./ynh-dev use-git yunohost moulinette
yunohost tools postinstall
```

Rum migration script:
```bash
yunohost tools migrations migrate -t 2
```
---

### Tests
- [ ] Fresh install works fine with install_script
- [x] migration from metronome to prosody (roster (contact list) is saved?)
- [ ] connection from clients:
  - [ ] packaged apps Movim, and Jappix
  - [x] external XMPP clients: done with Poezio locally
- [x] Regen-conf
- [x] backup/restore

---
### Release plan
- Create unstable `yunohost` debian package to test if everything works fine.
- migration script must be runned after package upgrade: how?
  - from debian script "debian/postinst": "yunohost tools migrations migrate -t 2"

---

### Extra
- [Interesting article about XMPP](https://www.erlang-solutions.com/blog/21-xmpp-use-cases-and-the-best-ways-to-achieve-them.html)
- [(fr) Interesting series articles](https://linuxfr.org/tags/parlons_xmpp/public)
- [Crowdfunding using Ğ1 libre currency](https://gannonce.duniter.org/#/announce/762de1a1-aa9a-4fa2-88de-0bcd12811215) · [D*](https://framasphere.org/posts/4032538)
- [Features of Prosody v.0.10 release](https://hg.prosody.im/0.10/file/tip/CHANGES)
- [Prosody 0.10.0 released](https://blog.prosody.im/prosody-0-10-0-released/)

### CODE REVIEW

**Comment by zamentur** [comment]
About using Prosody repo, we could import the .deb inside ou yunohost repo, no?

**Comment by Psycojoker** [comment]
Yes, we can theoretically do that https://github.com/YunoHost/doc/blob/master/build_packages.md#add-a-package-to-a-repository-manually

**Comment by polytan02** [comment]
I'm OK to run tests on my test server. Could you tel me how to migrate please ?

**Comment by M5oul** [comment]
> I'm OK to run tests on my test server. Could you tel me how to migrate please ?

Many works is still needed to be done before testing.

File: data/hooks/restore/26-conf_xmpp [hunk 1]
```diff
@@ -1,7 +1,7 @@
 backup_dir="$1/conf/xmpp"
 
-sudo cp -a $backup_dir/etc/. /etc/metronome
-sudo cp -a $backup_dir/var/. /var/lib/metronome
+sudo cp -a $backup_dir/etc/. /etc/prosody
+usudo cp -a $backup_dir/var/. /var/lib/prosody
```
Comment (mbugeia) [inline data/hooks/restore/26-conf_xmpp]: Typo here

Comment (M5oul) [inline data/hooks/restore/26-conf_xmpp]: Thanks.

**Comment by M5oul** [comment]
I rewrite the history for more clarity, review, less commits, beautiful history.

**Comment by Neustradamus** [comment]
Please update for stable 0.10.x ^^

**Comment by opi** [comment]
@Neustradamus we will use the current debian version, probably.

**Comment by M5oul** [comment]
We must target new version 0.10.x for sure

**Comment by M5oul** [comment]
serveimage

## Can I haz review?

**Comment by alexAubin** [comment]
Emaged, is it done ? Is this realf life ?

**Comment by Rafi594** [comment]
:+1:

**Comment by matlag** [comment]
Maybe I'll risk my life asking, but so'll be it :)

Can we (well, actually "you" would be more appropriate since in practice I'm not doing anything...) consider the SàT PubSub component instead of the Prosody one?
It should run fine with Prosody (that's actually the test reference), and seems much more mature and complete than the Prosody implementation.
For example:

- MAM for pubsub
- RSM (Manage large sets of results from queries)
- Other standard and experimental features

If this is implemented now, that would be one less data migration to care about while leaving the option to have Libervia (SàT web client) packaged later.

File: data/templates/prosody/prosody.cfg.lua [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		"ping"; -- Replies to XMPP pings with pongs
+		"register"; -- Allow users to register on this server using a client and change passwords
+		--"mam"; -- Store messages in an archive and allow users to access it
+
+	-- Admin interfaces
+		"admin_adhoc"; -- Allows administration via an XMPP client that supports ad-hoc commands
+		--"admin_telnet"; -- Opens telnet console interface on localhost port 5582
+
+	-- HTTP modules
+		--"bosh"; -- Enable BOSH clients, aka "Jabber over HTTP"
```
Comment (zamentur) [inline data/templates/prosody/prosody.cfg.lua]: In metronome config it was activated, it allows some web chat. Is there a reason to disable it ?

File: src/yunohost/data_migrations/0002_migration_from_metronome_to_prosody.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+from yunohost.service import service_regen_conf
+from moulinette.utils.filesystem import write_to_file, chown, rm
+
+
+class MyMigration(Migration):
+    def forward(self):
+
+        # Add Prosody repository and install it
+        write_to_file("/etc/apt/sources.list.d/prosody.list", "deb http://packages.prosody.im/debian jessie main")
+        subprocess.call("wget https://prosody.im/files/prosody-debian-packages.key -O- | sudo apt-key add -", shell=True)
```
Comment (zamentur) [inline src/yunohost/data_migrations/0002_migration_from_metronome_to_prosody.py]: Move the prosody package in our repo

**Comment by Neustradamus** [comment]
Metronome is alive, there is a 3.8 pre 1 since some days:
- https://metronome.im/
- https://metronome.im/changelogs/3_8
- https://github.com/maranda/metronome

cc @maranda :)

**Comment by Jaxom99** [comment]
hey @M5oul , is this still in the works ? linking from https://gannonce.duniter.org/#/announce/762de1a1-aa9a-4fa2-88de-0bcd12811215, i may donate a few G1 as well...

**Comment by M5oul** [comment]
Hey @Jaxom99!

Everything concerning XMPP is done. There is just the need to move from metronome to prosody left from a Debian dependencies’ migration management perspective. I am stuck to this final step.

This work were motivated because Metronome were no more maintained and developed for more than one year. Since then, Metronome development restarted.

For now, YunoHost project is still continuing with Metronome because it was easy to upgrade.
So, YunoHost v3.0 is now shipped with v3.9 of Metronome.

Then, If YunoHost project still wants to move to Prosody, this PR just needs to be finalized.

Concerning donation. It is as you want. The money collected for this crowdfunding will go in my pocket.
YunoHost project may officially accept Ğ1 currency in general. Let’s see…

**Comment by Jaxom99** [comment]
Thanks @M5oul for answering so quickly. It's nice to see that the project is still moving forward 😄 
Then my donation will be just a small compensation for all the work you've already put through !
And I'll gladly welcome a Yunohost Ğ1 account, although I already donate €€ via Liberapay.

**Comment by Kcchouette** [comment]
Hello.
Is this PR still active?
Thank you for this merge :)

**Comment by matlag** [comment]
Hi all!
If I trust Github and the second Git repo pointed at on Metronome website, there was not a single commit on Metronome for the past 5 months.
Is it time to put back this migration on the table?

**Comment by maranda** [comment]
Hi @matlag

There were no commits in 5 months because the code is stable and also there are *no feature* requests whatsoever as I explained over and over, the next thing on the roadmap is MIX but I'm waiting on implementation when there's anything supporting it that isnt just experimental.

You can post issues should you find bugs... 

And if you wish something to be added you can freely post a feature request *or* contribute yourself, otherwise staring at the commit list to see if a project is active or not is firmly pointless.

**Comment by trogeat** [comment]
Hi there,

I'v been investigating around why https://github.com/YunoHost-Apps/jitsi_ynh plugin didn't work and my brief conclusion is that there might be some feature missing in metronome.
Jitsi has been actively developed with Prosody xmpp server and modules, therefor metronome might have some missing stuff around Jingle protocol and don't handle it well.

I'm proposing to contribute to this PR to migrate to Prosody if you are ok with this in order to have a compatibility with jitsi_ynh.

We can also imagine that the xmpp service could be optional in yunohost at installation.

Would you mind if we also remove the strong dependancy between package yunohost and metronome to allow users to install another Yunohost-App with the xmpp server of his choice ?

Let me know if you are ok or not with this point of view.

Cheers !

**Pull request closed without merging**